STORM-3274: Migrates storm CLI to using argparse making documentation…

govind-menon opened the pull request:
… more accessible

govind-menon commented:
Tests will be forthcoming

srdo commented on storm-client/pom.xml:
```diff
@@ -240,6 +240,29 @@
+            <executions>
+                <execution>
+                    <configuration>
+                        <executable>python2.7</executable>
```
I'd like the storm-client build to still work on Windows. One option would be to only run this execution when the OS is UNIX-y, e.g.
```
<profile>
    <activation>
        <os><family>unix</family></os>
    </activation>
    <plugin-config-here/>
</profile>
```

srdo replied:
Or even better, set the executable name to python2.7 on Linux, and python on Windows.

govind-menon replied:
@srdo I've added a profile that would execute only on Unix - I didn't add one for Windows because that would change the file path separators that we're asserting on. I can change the test to make that configureable if you want

srdo replied:
Thanks, it looks good. I'm wondering if we can get away with also doing `pip install mock --user` as part of the Maven run, or is that a bad idea?

govind-menon replied:
@srdo It's a little messier to do that and better to leave it up to the environment to do it's own test setup. --user is for then there's no virtualenv (which will vary from machine to machine) - which is a good example of how the host itself should do the setup.

srdo commented on bin/storm.py:
```diff
@@ -132,13 +56,8 @@ def get_jars_full(adir):
-    return ret
+    return [os.path.join(adir, f) for f in files if f.endswith(".jar")]
 
-# If given path is a dir, make it a wildcard so the JVM will include all JARs in the directory.
```
Nit: Is this comment no longer relevant?

srdo commented on bin/storm.py:
```diff
@@ -156,49 +95,92 @@ def get_classpath(extrajars, daemon=True, client=False):
-    p = sub.Popen(command, stdout=sub.PIPE)
-    output, errors = p.communicate()
-    # python 3
-    if not isinstance(output, str):
```
Is this check not necessary for Python 3 anymore?

d2r replied:
Has this question been addressed?

d2r replied:
Yes this has been resolved already by keeping the needed checks.

srdo commented on bin/storm.py:
```diff
@@ -296,787 +239,1044 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+
+    sub_parser.add_argument("sql_file", metavar="sql-file")
+    sub_parser.add_argument(
+        "topology_name", metavar="topology-name", help="should be --explain to activate explain mode"
```
Nit: Can we use an argument group to do this instead of putting --explain into topology-name? https://docs.python.org/3/library/argparse.html#mutual-exclusion

govind-menon replied:
There are positional arguments added on L520 that don't allow mutual exclusion

srdo commented:
This is in reply to https://github.com/apache/storm/pull/2930#discussion_r245726485, for some reason Github won't let me put another comment there.

Makes sense. I think we should at least note that `mock` is a prerequisite in https://github.com/apache/storm/blob/master/DEVELOPER.md#prerequisites. Maybe add a line that users should install `pip` and then do `pip install mock --user`.

srdo commented:
Would it make sense to run the tests for Python 3 as well as Python 2.7?

srdo commented:
When I run the build, a storm.pyc file is generated in the storm/bin directory. This file ends up in the storm-dist distribution. I'm guessing this isn't intentional?

srdo commented:
Can we make running `storm.py` with 0 arguments equivalent to running `storm.py -h`? Running just `storm.py` gives the following output, which isn't very nice:

```
PS E:\apache-storm-2.0.1-SNAPSHOT\bin> ./storm
usage: storm [-h] [--config CONFIG] [-c C]
             {jar,localconfvalue,remoteconfvalue,local,sql,kill,upload-credentials,blobstore,heartbeats,activate,list,deactivate,rebalance,get-errors,kill_workers,admin,shell,repl,nimbus,pacemaker,supervisor,ui,logviewer,drpc-client,drpc,dev-zookeeper,version,classpath,server_classpath,monitor}
             ...
storm: error: too few arguments
```

srdo commented:
Nit: The existing help message prints the commands alphabetically. We might want to do the same with the new help message. Quick google suggests it is possible https://stackoverflow.com/questions/12268602/sort-argparse-help-alphabetically.

srdo commented on bin/storm.py:
```diff
@@ -296,787 +239,1044 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+
+def initialize_main_command():
+    main_parser = argparse.ArgumentParser(prog="storm")
+    main_parser.add_argument("--config", default=None, help="Override default storm conf")
```
Nit: It isn't obvious from the description what the difference between `-c` and `--config` is, or how to use them. Can we update the descriptions? For `--config` something like "Override default storm conf file", and for `-c` something like "Override storm conf properties, e.g. key=val,key2=val2".

govind-menon commented:
@srdo 

Addressed most of your comments

- Tests are run for python2 and python 3
- Help is sorted alphabetically now
- running just storm prints the entire help
- The mutual exclusion change isn't doable if there are other positional arguments (which there are for that command)

srdo commented:
Thanks. It looks great. A few final things:

* When running `mvn clean install -DskipTests`, the tests for the script are still run. Can we skip the executions if `skipTests` is true?
* Running the tests still generates a .pyc file in storm/bin, which ends up in the distribution. I think we should either remove it after the tests run, or explicitly exclude it from the storm-dist build.
* The help is sorted alphabetically, which is good. Can we do the same for the positional arguments (e.g. jar, logviewer, kill)?
* I would still like us to describe the new dependency on `mock` in DEVELOPER.md, as mentioned here https://github.com/apache/storm/pull/2930#issuecomment-452016830. I think it is ideal if people can blindly follow the setup guide and build Storm, having the build fail because there are extra setup steps that aren't mentioned in the guide could be demotivating to new people.

srdo commented:
The .pyc issue can be solved by adding `<PYTHONDONTWRITEBYTECODE>true</PYTHONDONTWRITEBYTECODE>` to the environment variables for the two `exec-maven-plugin` executions

govind-menon commented:
@srdo I've addressed the 4 comments you had - thank you for the review

kishorvpatil commented:
@govind-menon The `node-health-check` command  is missing from `storm.py` after this change.

govind-menon commented:
@kishorvpatil Fixed + added a test for it

kishorvpatil commented:
@govind-menon 
The CLI is not recognizing -c options are configuration parameters.

kishorvpatil commented on bin/storm.py:
```diff
@@ -234,60 +228,21 @@ def resolve_dependencies(artifacts, artifact_repositories, maven_local_repos_dir
-        "-Dstorm.conf.file=" + CONFFILE,
-        "-cp", get_classpath(extrajars, daemon, client=client),
-    ] + jvmopts + [klass] + list(args)
+                   JAVA_CMD, jvmtype,
```
excess tabulation?

govind-menon replied:
No I was formatting it so it was more readable

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+    zk_cli_parser.add_argument(
+        "-t", "--time-out", default=None, help="""Set the timeout to use, defaults to storm
+            zookeeper timeout."""
+    )
```
* `type=check_positive`?
* Does this have to be positive? What happens on `0`?
* What are the units?

d2r replied:
This has been addressed.

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+    supervision with a tool like daemontools or monit.
+
+    See Setting up a Storm cluster for more information.
+    (http://storm.apache.org/documentation/Setting-up-a-Storm-cluster
```
Missing a closing `)` parenthesis

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+    sub_parser = subparsers.add_parser("drpc-client", help=command_help, formatter_class=SortingHelpFormatter)
+
+    sub_parser.add_argument(
+        "-f", "--function", default=None, help="""If a -f argument is supplied to set the function name all of the arguments are treated
```
Suggest `If the -f argument`...

d2r commented on bin/storm.py:
```diff
@@ -156,49 +107,92 @@ def get_classpath(extrajars, daemon=True, client=False):
+        print("*" * 20)
+        print('''The storm client can only be run from within a release. 
+You appear to be trying to run the client from a checkout of Storm's source code.
+You can download a Storm release at http://storm.apache.org/downloads.html")''')
```
All of the links can be updated to use `https`.

d2r replied:
This has been addressed.

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+        "-w", "--wait-time-secs",
+        help="""override the length of time Storm waits between deactivation and shutdown""",
+        default=None
+    )
```
Do we want a `check_nonnegative` function?

d2r replied:
This has been addressed.

govind-menon commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+    )
+
+def add_topology_jar_options(parser):
+    parser.add_argument("topology_jar_path", help="will upload the jar at topology-jar-path when the topology is submitted.")
```
I didn't understand this part - they're positional arguments so it doesn't change the invocation?

d2r replied:
Yes this was fixed.

govind-menon commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+        "activate", help="Activates the specified topology's spouts.", formatter_class=SortingHelpFormatter
+    )
+
+    sub_parser.add_argument("topology-name")
```
The java command only takes one topology

d2r replied:
Yes, this is correct. I misread the current version.

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+    initialize_version_subcommand(subparsers)
+    initialize_classpath_subcommand(subparsers)
+    initialize_server_classpath_subcommand(subparsers)
+    initialize_monitor_subcommand(subparsers)
```
This has been addressed.

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+        help="lets you turn on java debugging and set the parameters passed to -agentlib:jdwp on the JDK" +
+             "e.g transport=dt_socket,address=localhost:8000 will open up a debugging server on port 8000",
+        default=None
+    )
```
This has been addressed.

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+    sub_parser.add_argument(
+        "-n", "--num-workers", default=None,
+        help="change the number of requested workers"
+    )
```
This has been addressed.

govind-menon commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+    sub_parser.add_argument(
+        "-e", "--executors", action="append", default=[],
+        help="change the number of executors for a given component"
+    )
```
I added the validation for this in the method itself. Also the default=[] is equivalent to nargs='*'

d2r replied:
Ok sounds good.

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+    replication_parser.add_argument(
+        "--update", action="store_true", help=" It is used to update the replication factor of a blob."
+    )
+    replication_parser.add_argument("-r", "--replication-factor", default=None)
```
This has been addressed.

d2r commented on bin/storm.py:
```diff
@@ -296,787 +251,1090 @@ def exec_storm_class(klass, jvmtype="-server", jvmopts=[], extrajars=[], args=[]
+
+    sub_parser.add_argument("sql_file", metavar="sql-file")
+    sub_parser.add_argument(
+        "topology_name", metavar="topology-name", help="should be --explain to activate explain mode"
```
This has been addressed by implementing the mutually exclusive groups.
